pyccel/codegen/printing/cwrappercode.py

```diff
@@ -0,0 +1,385 @@
+            return requested_name
+        else:
+            incremented_name, _ = create_incremented_string(used_names, prefix=requested_name)
+            return incremented_name
+
+    def get_new_PyObject(self, name, used_names):
+        return Variable(dtype=PyccelPyObject(),
+                        name=self.get_new_name(used_names, name),
+                        is_pointer=True)
+
+    def find_in_dtype_registry(self, dtype, prec):
+        try :
+            return dtype_registry[(dtype, prec)]
+        except KeyError:
+            return CCodePrinter.find_in_dtype_registry(self, dtype, prec)
+
+    def get_cast_function_call(self, cast_type, arg):
+        """
+        Represents a call to cast function responsible of the conversion of one data type into another.
+
+        Parameters:
+        ----------
+        used_names: list of strings
+            List of variable and function names
+        cast_type: string
+            The type of cast function on format 'data type_to_data type'
+        from_variable: variable
+            the variable needed to cast
+        to_variable: variable
+            the result of the cast operation
+        """
+
+        if cast_type in self._cast_functions_dict:
+            cast_function = self._cast_functions_dict[cast_type]
+
+        else:
+            cast_function_name = self.get_new_name(self._global_names, cast_type)
+
+            cast_function = cast_function_registry[cast_type](cast_function_name)
+
+            self._cast_functions_dict[cast_type] = cast_function
+
+        return FunctionCall(cast_function, [arg])
+
+    def get_PyArgParseType(self, used_names, variable):
+        """
+        Responsible of collecting the python variable into which the result will be collected and the needed cast function
+
+        Parameters:
+        ----------
+        used_names : list of strings
+            List of variable and function names
+        variable : variable
+        """
+
+        if variable.dtype is NativeBool():
+            collect_type = NativeInteger()
+            collect_var = Variable(dtype=collect_type, precision=4,
+                name = self.get_new_name(used_names, variable.name+"_tmp"))
+            cast_function = self.get_cast_function_call('pyint_to_bool', collect_var)
+            body = [Assign(variable, cast_function)]
+            return collect_var, body
+
+        if variable.dtype is NativeComplex():
+            collect_type = PyccelPyObject()
+            collect_var = Variable(dtype=collect_type, is_pointer=True,
+                name = self.get_new_name(used_names, variable.name+"_tmp"))
+            cast_function = self.get_cast_function_call('pycomplex_to_complex', collect_var)
+            body = [Assign(variable, cast_function)]
+            if isinstance(variable, ValuedVariable):
+                # Decrement PyObject counter which was incremented by PyComplex_FromDoubles
+                body += [FunctionCall(Py_DECREF, [collect_var])]
```

Comment:
This leads to the argument being derefenced if it is provided which leads to a segmentation fault

Reply:
This has been solved by initialising the PyObject to None and adding an `If`` statement to test whether PyObject points to an argument or not. If it does the argument is collected. If not the default argument is used.

This also reduces the number of casts necessary